[docs] Handbook pages (#1141)
mui/base-ui: docs/src/app/(public)/(content)/react/handbook/composition/page.mdx

# Composition

<Subtitle>A guide to composing Base UI components with your own React components.</Subtitle>
<Meta
  name="description"
  content="A guide to composing Base UI components with your own React components."
/>

### Composing custom React components

The `render` prop lets you override a component's rendered element with your own React component. For example, `Menu.Trigger` renders a `<button>` by default. The code snippet below shows how to use a custom component instead.

```tsx title="index.tsx"

> <Menu.Trigger render={<MyFancyButton size="md" />} />

Reviewer: Children can (and should, for simplicity) be placed as the original component children, not the overridden one.

```

> The custom component must forward a `ref`, and spread all the received props, to its underlying DOM node.

Reviewer 2: There's only one ref (the ref) and the props are either passed to a node or spread on (onto?) a node

### Composing multiple components

In situations where you need to compose multiple Base UI components with custom React components, `render` props can be nested as deeply as necessary. Working with Tooltip is a common example.

```tsx title="index.tsx"

> <Dialog.Root>
>   <Tooltip.Root>
>     <Tooltip.Trigger
>     {/* prettier-ignore */}
>       render={
>         <Dialog.Trigger
>           render={
>             <MyFancyButton size="md" />
>           }
>         />
>       }
>     />
>     <Tooltip.Portal>...</Tooltip.Portal>
>   </Tooltip.Root>
>   <Dialog.Portal>...</Dialog.Portal>
> </Dialog.Root>

Reviewer 2: imo the root parts make the example a bit more contrived
```suggestion
<Tooltip.Trigger
  render={
    <Dialog.Trigger
      render={<MyFancyButton size="md">Open dialog</MyButton>}
    />
  }
/>
```

Author: Gonna leave this alone because it would help some people, myself included.

Reviewer 2: All right let's just do
```
render={<MyButton size="md">Open dialog</MyButton>}
```
within

```

> ### Changing the default rendered element
> 
> You can also use the `render` prop to override the rendered element of the component. For example, `Menu.Item` renders a `<div>` by default.
> The code snippet below shows how to render an `<a>` instead.

Reviewer 2: ```suggestion

 ### Changing the default rendered element

 You can also use the `render` prop to override the rendered element of the component.
 
 For example, you may render a menu item as an `<a>` element so that it works like a link.
```

```tsx title="index.tsx"
import * as React from 'react';
import { Menu } from '@base-ui-components/react/menu';

export default () => (
  <Menu.Root>
    <Menu.Trigger>Song</Menu.Trigger>
    <Menu.Portal>
      <Menu.Positioner>
        <Menu.Popup>
          <Menu.Item render={<a href="base-ui.com" />}>Add to Library</Menu.Item>
        </Menu.Popup>
      </Menu.Positioner>
    </Menu.Portal>
  </Menu.Root>
);
```

> Changing the default rendered element can break important accessibility features like focus management, screen reader announcements, and keyboard events. Each Base UI component renders the most appropriate element by default, and in most cases, rendering a different element is not recommended.

Reviewer 2: I don't think we need to scare people away from the render prop here, since most components implement the right semantics using ARIA roles and keyboard accessibility doesn't hinge on the tag either. Composing with custom react components is much more likely to break something

```suggestion
Each Base UI component renders the most appropriate element by default, and in most cases, rendering a different element is recommended only on a case-by-case basis.
```

Reviewer: I agree with the original suggestion more. I don't think we should encourage anyone to change the rendered element. We can save ourselves some trouble when the issues that something exotic doesn't work start coming.

### Function render prop

If you are working in an extremely performance-sensitive application, you might want to use the function `render` prop instead of the element `render` prop.

```tsx title="switch.tsx"
<Switch.Thumb
  render={(props, internalState) => (
    <span {...props}>{internalState.checked ? <CheckedIcon /> : <UncheckedIcon />}</span>
  )}
/>
```

> Using a function gives you complete control over spreading props and also allows you to render different content based on the component's state.

Reviewer 2: - The canonical name of the feature is "Render function"
- Should be just `state`, not `internalState` imo

Reviewer: I'd reorder these paragraphs. IMO, the primary reason for using the render function is having control over prop spreading (so you can render a tree of components for example: `(props) => <div {...props}><div>{props.children}</div></div>`, then being able to render content conditionally, then performance.

# Composition

<Subtitle>A guide to composing Base UI components with your own React components.</Subtitle>
<Meta
  name="description"
  content="A guide to composing Base UI components with your own React components."
/>

### Composing custom React components

Use the `render` prop to compose a Base UI part with your own React components.

For example, most triggers render a `<button>` by default. The code snippet below shows how to use a custom button instead.

```tsx title="index.tsx"

> <Menu.Trigger render={<MyButton size="md">Open menu</MyButton>} />

Reviewer: Looks like my comment was lost (GH is having some problems)
We should recommend placing children in the original component not the one in the render prop for simplicity (similarily to Ariakit: https://ariakit.org/guide/composition#merging-the-rendered-element-props, Flat is better than nested)

Reviewer 2: OK @colmtuite don't listen to me, listen to @michaldudak 😅

    {/* prettier-ignore */}
      render={
        <Dialog.Trigger
          render={
            <MyButton size="md">Open dialog</MyButton>
          }
        />
      }
    />
    <Tooltip.Portal>...</Tooltip.Portal>
  </Tooltip.Root>
  <Dialog.Portal>...</Dialog.Portal>
</Dialog.Root>
```

### Changing the default rendered element

You can also use the `render` prop to override the rendered element of the component.

For example, `Menu.Item` renders a `<div>` by default. The code snippet below shows how to render it as an `<a>` element so that it works like a link.

```tsx title="index.tsx"
import * as React from 'react';
import { Menu } from '@base-ui-components/react/menu';

export default () => (
  <Menu.Root>
    <Menu.Trigger>Song</Menu.Trigger>
    <Menu.Portal>
      <Menu.Positioner>
        <Menu.Popup>
          <Menu.Item render={<a href="base-ui.com" />}>Add to Library</Menu.Item>
        </Menu.Popup>
      </Menu.Positioner>
    </Menu.Portal>
  </Menu.Root>
);
```

Each Base UI component renders the most appropriate element by default, and in most cases, rendering a different element is recommended only on a case-by-case basis.

### Render function

> If you are working in an extremely performance-sensitive application, you might want to use the function `render` prop instead of the element `render` prop.

Reviewer 2: This doesn't compute for me

```suggestion
If you are working in an extremely performance-sensitive application, you might want to pass a function to the `render` prop instead of a React element.
```

Reviewer: https://github.com/mui/base-ui/pull/1141#discussion_r1888656196 (it's collapsed above, buried between other "outdated" comments)

import { Menu } from '@base-ui-components/react/menu';

export default () => (
  <Menu.Root>
    <Menu.Trigger>Song</Menu.Trigger>
    <Menu.Portal>
      <Menu.Positioner>
        <Menu.Popup>
          <Menu.Item render={<a href="base-ui.com" />}>Add to Library</Menu.Item>
        </Menu.Popup>
      </Menu.Positioner>
    </Menu.Portal>
  </Menu.Root>
);
```

Changing the default rendered element can break important accessibility features like focus management, screen reader announcements, and keyboard events. Each Base UI component renders the most appropriate element by default, and in most cases, rendering a different element is not recommended.

### Function render prop

If you are working in an extremely performance-sensitive application, you might want to use the function `render` prop instead of the element `render` prop.

```tsx title="switch.tsx"
<Switch.Thumb
  render={(props, internalState) => (
    <span {...props}>{internalState.checked ? <CheckedIcon /> : <UncheckedIcon />}</span>
  )}
/>
```

Using a function gives you complete control over spreading props and also allows you to render different content based on the component's state.

> ### Component state as an argument

Reviewer 3: This should be in the styling page, no?

Reviewer 3: Maybe move it to https://github.com/mui/base-ui/pull/1141/files#diff-cd91ca9130b97e2deaa53abbd220613432d1f76d9edc7fd5e26d00c46a3f4e31R15
